README.md
# datadog-events

[![npm](https://img.shields.io/npm/v/datadog-events.svg?style=for-the-badge)](https://www.npmjs.com/package/datadog-events) [![npm](https://img.shields.io/npm/dt/datadog-events.svg?style=for-the-badge)](https://www.npmjs.com/package/datadog-events) [![David](https://img.shields.io/david/KyleRoss/node-datadog-events.svg?style=for-the-badge)](https://david-dm.org/KyleRoss/node-datadog-events) [![Travis](https://img.shields.io/travis/KyleRoss/node-datadog-events/master.svg?style=for-the-badge)](https://travis-ci.org/KyleRoss/node-datadog-events) [![license](https://img.shields.io/github/license/KyleRoss/node-datadog-events.svg?style=for-the-badge)](https://github.com/KyleRoss/node-datadog-events/blob/master/LICENSE) [![Beerpay](https://img.shields.io/beerpay/KyleRoss/node-datadog-events.svg?style=for-the-badge)](https://beerpay.io/KyleRoss/node-datadog-events)


Send events to DataDog **without** DogStatsD or StatsD, using the standard DataDog API. Supports Node 6+.

## Getting Started
### Install
```sh
npm install datadog-events --save
```

### Use
By default, a function is exported that will create a new instance of `DataDogEvents` for you. You may either use the default or create your own instance of the class:

#### Default Exports
```js
const ddEvents = require('datadog-events')({ /* options */ });

async function doSomething() {
    // ...
    
    if(err) {
        await ddEvents.error('Some error!', 'These are some details');
    }
}
```

#### Create Instance
```js
const DataDogEvents = require('datadog-events').DataDogEvents;
const ddEvents = new DataDogEvents({ /* options */ });
```

## API Documentation
### Options

#### Global DataDogEvents Options
The following options are available when creating a new instance of `DataDogEvents` or when calling the default export function.

| Option         | Type          | Required? | Description                                                                                                             | Default                       |
|----------------|---------------|-----------|-------------------------------------------------------------------------------------------------------------------------|-------------------------------|
| apiKey         | String        | Yes       | Your DataDog API key. This may be populated using environment variable `DATADOG_API_KEY`.                               | `process.env.DATADOG_API_KEY` |
| domain         | String        | No        | DataDog domain to use for the API. Useful for switching to the EU version (ex. `datadoghq.eu`).                         | `"datadoghq.com"`             |
| titlePrefix    | String        | No        | Optional text to prefix all event titles with.                                                                          | `null`                        |
| bodyPrefix     | String        | No        | Optional text to prefix all event bodies with.                                                                          | `null`                        |
| bodyPostfix    | String        | No        | Optional test to postfix all event bodies with.                                                                         | `null`                        |
| priority       | String        | No        | Priority for all events. Can be either `normal` or `low`.                                                               | `"normal"`                    |
| host           | String        | No        | Optional host name to attach to all events.                                                                             | `null`                        |
| tags           | Array[String] | No        | Optional tags to attach to all events.                                                                                  | `[]`                          |
| aggregationKey | String        | No        | Optional key that will allow DataDog to aggregate all events under.                                                     | `null`                        |
| sourceType     | String        | No        | Optional source type name. See [here](https://docs.datadoghq.com/integrations/faq/list-of-api-source-attribute-value/). | `null`                        |
| markdown       | Boolean       | No        | Format all event bodies as markdown.                                                                                    | `true`                        |

#### Event Options
The following options may be provided when sending events with any of the methods documented below. Most of these options will override the global options listed above and are all optional.

| Option         | Type          | Description                                                                                                             |
| -------------- | ------------- | ----------------------------------------------------------------------------------------------------------------------- |
| date           | Date          | Date object of the event. By default, it will be the current date/time.                                                 |
| priority       | String        | Priority for the event. Can be either `normal` or `low`.                                                                |
| host           | String        | Optional host name to attach to the event.                                                                              |
| tags           | Array[String] | Tags to append to the event. Tags provided will be appended to the global `options.tags`.                               |
| aggregationKey | String        | Optional key that will allow DataDog to aggregate all events under.                                                     |
| sourceType     | String        | Optional source type name. See [here](https://docs.datadoghq.com/integrations/faq/list-of-api-source-attribute-value/). |
| markdown       | Boolean       | Format the body as markdown.                                                                                            |

### DataDogEvents Class
The main class for `datadog-events`.

#### constructor([options={}])
The constructor for `DataDogEvents`.

| Argument | Type    | Required? | Description                                                     |
| -------- | ------- | --------- | --------------------------------------------------------------- |
| options  | ?Object | No        | [Global options](#global-datadogevents-options) for the class.  |

##### Example
```js
// Shortcut
const ddEvents = require('datadog-events')({ /* options */ });

// Create your own instance
const DataDogEvents = require('datadog-events').DataDogEvents;
const ddEvents = new DataDogEvents({ /* options */ });
```

###### Returns - _DataDogEvents_
> Returns instance of DataDogEvents.

### Methods

#### ddEvents.<error|warning|info|success>(title, body[, options={}])
Shortcut methods for `ddEvents.sendEvent()`. Will send an event with the given type of the method.

| Argument | Type                | Required? | Description                                                                                                                                                       |
| -------- | ------------------- | --------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| title    | String              | Yes       | Title for the event.                                                                                                                                              |
| body     | String|Object|Error | Yes       | The body of the event. If `options.markdown` is `true`, it will be formatted as markdown. If an Object or Error is passed in, it will be formatted into markdown. |
| options  | ?Object             | No        | Optional [event options](#event-options).                                                                                                                         |

##### Example
```js
ddEvents.error('There was an error!', error)
    .then(response => console.log(response));

await ddEvents.success('Completed a process!', '**The process was completed!**');
```
###### Returns - _Promise[Object]_
> Promise resolves with object returned from the DataDog API containing `status` and `event` keys. Rejects with error if failed to send.

#### ddEvents.sendEvent(type, title, body[, options={}])
Sends an event to DataDog.

| Argument | Type                | Required? | Description                                                                                                                                                       |
| -------- | ------------------- | --------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| type     | String              | Yes       | Alert type of the event. Can be `error`, `warning`, `info` or `success`.                                                                                          |
| title    | String              | Yes       | Title for the event.                                                                                                                                              |
| body     | String|Object|Error | Yes       | The body of the event. If `options.markdown` is `true`, it will be formatted as markdown. If an Object or Error is passed in, it will be formatted into markdown. |
| options  | ?Object             | No        | Optional [event options](#event-options).                                                                                                                         |

##### Example
```js
ddEvents.sendEvent('warning', 'Something happened', '**Warning:** Something crazy happened')
    .then(resp => console.log(resp));

await ddEvents.sendEvent('info', 'Results from my process', resultsObject);
```

###### Returns - _Promise[Object]_
> Promise resolves with object returned from the DataDog API containing `status` and `event` keys. Rejects with error if failed to send.


### Properties
#### ddEvents.options - _Object_
The [global options](#global-datadogevents-options) for the current instance of DataDogEvents. You may change any of the options at any time.

### Environment Variables
| Environment Variable | Type   | Option           | Description                                 |
|----------------------|--------|------------------|---------------------------------------------|
| DATADOG_API_KEY      | String | `options.apiKey` | Sets the API Key for DataDog automatically. |
| DATADOG_DOMAIN       | String | `options.domain` | Sets the domain for the DataDog API.        |

## Tests
Tests run automatically in Travis, although you may run them on your own machine. If you do, you must have your own DataDog account and API key. Make sure you add the `DATADOG_API_KEY` environment variable before running the tests.

```sh
DATADOG_API_KEY=MY_API_KEY npm run test
```

## License
MIT License. See [License](https://github.com/KyleRoss/node-datadog-events/blob/master/LICENSE) in the repository.


HISTORY.md
# datadog-events History

## 1.1.0 (4/8/2019)
* Added ability to switch the DataDog domain for the API. (#1)

## 1.0.0 (5/30/2018)
* Initial Release


test.js
const assert = require('chai').assert;
const datadogEvents = require('./');

describe ('Exports', () => {
    it('should export a function', () => {
        assert.isFunction(datadogEvents);
    });
    
    it ('should have access to the DataDogEvents class', () => {
        assert.isFunction(datadogEvents.DataDogEvents);
    });
    
    it ('should create a new instance', () => {
        let events = datadogEvents();
        assert.instanceOf(events, datadogEvents.DataDogEvents);
    });
    
    it ('should allow setting custom options', () => {
        let events = datadogEvents({
            titlePrefix: 'test'
        });
        
        assert.equal(events.options.titlePrefix, 'test');
    });
});

describe ('DataDogEvents', () => {
    describe('Default Options', () => {
        let events = datadogEvents();
        
        it('should should have all default options', () => {
            assert.hasAllKeys(events.options, [
                'apiKey',
                'titlePrefix',
                'bodyPrefix',
                'bodyPostfix',
                'priority',
                'host',
                'tags',
                'aggregationKey',
                'sourceType',
                'markdown'
            ]);
        });
        
        it('should set apiKey to evnvironment variable', () => {
            assert.isString(events.options.apiKey);
        });
        
        it ('should set all default options', () => {
            assert.equal(events.options.titlePrefix, null);
            assert.equal(events.options.bodyPrefix, null);
            assert.equal(events.options.bodyPostfix, null);
            assert.equal(events.options.priority, 'normal');
            assert.equal(events.options.host, null);
            assert.isArray(events.options.tags);
            assert.equal(events.options.aggregationKey, null);
            assert.equal(events.options.sourceType, null);
            assert.equal(events.options.markdown, true);
        });
    });
    
    describe ('Custom Options', () => {
        let events = datadogEvents({
            apiKey: 'test',
            titlePrefix: 'test',
            bodyPrefix: 'test',
            bodyPostfix: 'test',
            priority: 'low',
            host: 'test.test',
            tags: ['app:test'],
            aggregationKey: '1234567890',
            sourceType: 'test'
        });
        
        it('should set the custom options correctly', () => {
            assert.equal(events.options.apiKey, 'test');
            assert.equal(events.options.titlePrefix, 'test');
            assert.equal(events.options.bodyPrefix, 'test');
            assert.equal(events.options.bodyPostfix, 'test');
            assert.equal(events.options.priority, 'low');
            assert.equal(events.options.host, 'test.test');
            assert.deepEqual(events.options.tags, ['app:test']);
            assert.equal(events.options.aggregationKey, '1234567890');
            assert.equal(events.options.sourceType, 'test');
            assert.equal(events.options.markdown, true);
        });
        
        it ('should allow setting options after instantiation', () => {
            events.options.priority = 'normal';
            events.options.tags.push('my-test-app');
            
            assert.equal(events.options.priority, 'normal');
            assert.deepEqual(events.options.tags, ['app:test', 'my-test-app']);
        });
    });
    
    describe ('Constructor Errors', () => {
        it('should throw error if API key is not set', () => {
            assert.throws(datadogEvents.bind(null, { apiKey: null }), Error);
        });
    });
    
    describe ('Methods', () => {
        let events = datadogEvents();
        
        beforeEach(() => {
            events.options = {
                apiKey: process.env.DATADOG_API_KEY,
                titlePrefix: null,
                bodyPrefix: null,
                bodyPostfix: null,
                priority: 'normal',
                host: null,
                tags: [],
                aggregationKey: null,
                sourceType: null,
                markdown: true
            };
        });
        
        describe('_prepareParams()', () => {
            it ('should have method', () => {
                assert.isFunction(events._prepareParams);
            });
            
            it ('should return object with correct keys', () => {
                let params = events._prepareParams({});
                assert.isObject(params);
                assert.hasAllKeys(params, [
                    'title',
                    'text',
                    'priority',
                    'tags',
                    'alert_type'
                ]);
            });
            
            it ('should return custom options when provided', () => {
                let params = events._prepareParams({
                    title: 'test',
                    body: 'test',
                    priority: 'low',
                    tags: ['app:test'],
                    type: 'warning',
                    date: new Date(),
                    host: 'test.test',
                    aggregationKey: '1234567890',
                    sourceType: 'test'
                });
                
                assert.hasAllKeys(params, [
                    'title',
                    'text',
                    'priority',
                    'tags',
                    'alert_type',
                    'date_happened',
                    'host',
                    'aggregation_key',
                    'source_type_name'
                ]);
            });
            
            it ('should merge tags', () => {
                events.options.tags.push('app:test');
                let params = events._prepareParams({
                    tags: ['my-test-tag']
                });
                
                assert.deepEqual(params.tags, ['app:test', 'my-test-tag']);
            });
            
            it ('should add markdown identifiers when options.markdown=true', () => {
                let params = events._prepareParams({
                    body: 'test'
                });
                
                assert.equal(params.text, '%%% \n test \n %%%');
            });
            
            it('should not add markdown identifiers when options.markdown=false', () => {
                events.options.markdown = false;
                let params = events._prepareParams({
                    body: 'test'
                });

                assert.equal(params.text, 'test');
            });
            
            it ('should markdownify errors', () => {
                let params = events._prepareParams({
                    body: new Error('test')
                });
                
                assert.match(params.text, /^%%%/i);
                assert.include(params.text, 'Error: test');
                assert.match(params.text, /%%%$/i);
            });
            
            it('should markdownify objects', () => {
                let params = events._prepareParams({
                    body: { test : 'test' }
                });

                assert.match(params.text, /^%%%/i);
                assert.include(params.text, '"test": "test"');
                assert.match(params.text, /%%%$/i);
            });
            
            it('should markdownify errors/objects even when options.markdown=false', () => {
                events.options.markdown = false;
                let params = events._prepareParams({
                    body: new Error('test')
                });

                assert.match(params.text, /^%%%/i);
                assert.match(params.text, /%%%$/i);
            });
            
            it ('should add prefixes and postfixes', () => {
                events.options.titlePrefix = 'test! ';
                events.options.bodyPrefix = 'body! ';
                events.options.bodyPostfix = ' !body';
                
                let params = events._prepareParams({
                    title: 'test',
                    body: 'body'
                });
                
                assert.equal(params.title, 'test! test');
                assert.equal(params.text, '%%% \n body! body !body \n %%%');
            });
        });
        
        describe ('sendEvent()', () => {
            it('should have method', () => {
                assert.isFunction(events.sendEvent);
            });
            
            it('should send an event', (done) => {
                events.sendEvent('info', 'test title', 'test body')
                    .then(res => {
                        assert.isObject(res);
                        assert.containsAllKeys(res, [
                            'status',
                            'event'
                        ]);
                        assert.equal(res.status, 'ok');
                        done();
                    })
                    .catch(done);
            });
            
            it ('should return error when API key is not provided', (done) => {
                events.options.apiKey = null;
                events.sendEvent('info', 'test title', 'test body')
                    .then(() => {
                        done(new Error('Should not return successful!'));
                    })
                    .catch(e => {
                        assert.instanceOf(e, Error);
                        assert.equal(e.message, 'Invalid API Key provided');
                        done();
                    });
            });
        });
        
        describe('error()', () => {
            it('should have method', () => {
                assert.isFunction(events.error);
            });
        });
        
        describe('warning()', () => {
            it('should have method', () => {
                assert.isFunction(events.warning);
            });
        });
        
        describe('info()', () => {
            it('should have method', () => {
                assert.isFunction(events.info);
            });
        });
        
        describe('success()', () => {
            it('should have method', () => {
                assert.isFunction(events.success);
            });
        });
    });
});


lib/DataDogEvents.js
"use strict";
const axios = require('axios');
const isError = require('is-error');


/**
 * @class DataDogEvents
 */
class DataDogEvents {
    /**
     * Creates an instance of DataDogEvents.
     * @param {?Object} [options={}] Options to configure DataDogEvents
     * 
     * @memberOf DataDogEvents
     */
    constructor(options = {}) {
        this.options = Object.assign({
            apiKey: process.env.DATADOG_API_KEY || null,
            titlePrefix: null,
            bodyPrefix: null,
            bodyPostfix: null,
            priority: 'normal',
            host: null,
            tags: [],
            aggregationKey: null,
            sourceType: null,
            markdown: true,
            silent: false
        }, options || {});
        
        if(!this.options.apiKey) 
            throw new Error('DataDog API key was not set');
        
        
        
        ['error', 'warning', 'info', 'success'].forEach(type => {
            this[type] = (title, body, options = {}) => {
                return this.sendEvent(type, title, body, options);
            };
        });
    }
    
    /**
     * Sends an event to DataDog.
     * 
     * @param {String}  type         The alert type to send (can be `error`, `warning`, `info` or `success`).
     * @param {String}  title        The title of the event.
     * @param {String}  body         The body of the event which may contain markdown.
     * @param {?Object} [options={}] Additional options to configure the event.
     * @returns {Promise}
     * 
     * @memberOf DataDogEvents
     */
    sendEvent(type, title, body, options = {}) {
        options = Object.assign(options || {}, { title, body, type });
        
        let data = this._prepareParams(options);
        
        return new Promise((resolve, reject) => {
            axios.request({
                method: 'post',
                url: 'https://app.datadoghq.com/api/v1/events',
                params: {
                    api_key: this.options.apiKey
                },
                data
            }).then(resp => {
                resolve(resp.data);
            }).catch(err => {
                let status = err.response.status;
                
                if(status === 403) {
                    err.message = 'Invalid API Key provided';
                } else {
                    let resp = err.response.data;
                    if(resp.errors) err.message = resp.errors.join(', ');
                }
                
                reject(err);
            });
        });
    }
    
    /**
     * Prepares parameters to match format of the DataDog API before sending an event.
     * 
     * @param   {any}    cfg  Custom options for the event.
     * @returns {Object}      The compiled params object.
     * 
     * @memberOf DataDogEvents
     */
    _prepareParams(cfg) {
        let opts = this.options;
        
        if(typeof cfg.body === 'object') {
            if(isError(cfg.body)) {
                cfg.body = [
                    '```',
                    cfg.body.toString(),
                    cfg.body.stack,
                    '```'
                ].join('\n');
            } else {
                cfg.body = [
                    '```',
                    JSON.stringify(cfg.body, null, 4),
                    '```'
                ].join('\n');
            }
            
            cfg.markdown = true;
        }
        
        let params = {
            title: `${opts.titlePrefix || ''}${cfg.title}`,
            text: `${opts.bodyPrefix || ''}${cfg.body}${opts.bodyPostfix || ''}`,
            priority: cfg.priority || opts.priority || 'normal',
            tags: opts.tags || [],
            alert_type: cfg.type || 'info'
        };
        
        if(cfg.markdown || opts.markdown)
            params.text = `%%% \n ${params.text} \n %%%`;
        
        if(cfg.date && cfg.date instanceof Date)
            params.date_happened = Math.round(cfg.date.getTime() / 1000);
        
        if(cfg.host || opts.host)
            params.host = cfg.host || opts.host;
        
        if(cfg.tags && Array.isArray(cfg.tags))
            params.tags = params.tags.concat(cfg.tags);
        
        if(cfg.aggregationKey || opts.aggregationKey)
            params.aggregation_key = cfg.aggregationKey || opts.aggregationKey;
        
        if(cfg.sourceType || opts.sourceType)
            params.source_type_name = cfg.sourceType || opts.sourceType;
        
        return params;
    }
}

module.exports = DataDogEvents;


index.js
/**
 * @module datadog-events
 * @description Send events to DataDog **without** DogStatsD or StatsD
 * @author Kyle Ross
 */
"use strict";
const axios = require('axios');
const isError = require('is-error');

/**
 * @class DataDogEvents
 */
class DataDogEvents {
    /**
     * Creates an instance of DataDogEvents.
     * @param {?Object} [options={}] Options to configure DataDogEvents
     * 
     * @memberOf DataDogEvents
     */
    constructor(options = {}) {
        this.options = Object.assign({
            apiKey: process.env.DATADOG_API_KEY || null,
            domain: process.env.DATADOG_DOMAIN || 'datadoghq.com',
            titlePrefix: null,
            bodyPrefix: null,
            bodyPostfix: null,
            priority: 'normal',
            host: null,
            tags: [],
            aggregationKey: null,
            sourceType: null,
            markdown: true
        }, options || {});
        
        if(!this.options.apiKey) 
            throw new Error('DataDog API key was not set');
        
        ['error', 'warning', 'info', 'success'].forEach(type => {
            this[type] = (title, body, options = {}) => {
                return this.sendEvent(type, title, body, options);
            };
        });
    }
    
    /**
     * Sends an event to DataDog.
     * 
     * @param {String}  type         The alert type to send (can be `error`, `warning`, `info` or `success`).
     * @param {String}  title        The title of the event.
     * @param {String}  body         The body of the event which may contain markdown.
     * @param {?Object} [options={}] Additional options to configure the event.
     * @returns {Promise}
     * 
     * @memberOf DataDogEvents
     */
    sendEvent(type, title, body, options = {}) {
        options = Object.assign(options || {}, { title, body, type });
        
        let data = this._prepareParams(options);
        
        return new Promise((resolve, reject) => {
            axios.request({
                method: 'post',
                url: `https://app.${this.options.domain}/api/v1/events`,
                params: {
                    api_key: this.options.apiKey
                },
                data
            }).then(resp => {
                resolve(resp.data);
            }).catch(err => {
                let status = err.response.status;
                
                if(status === 403) {
                    err.message = 'Invalid API Key provided';
                } else {
                    let resp = err.response.data;
                    if(resp.errors) err.message = resp.errors.join(', ');
                }
                
                reject(err);
            });
        });
    }
    
    /**
     * Prepares parameters to match format of the DataDog API before sending an event.
     * 
     * @param   {any}    cfg  Custom options for the event.
     * @returns {Object}      The compiled params object.
     * 
     * @memberOf DataDogEvents
     */
    _prepareParams(cfg) {
        let opts = this.options;
        
        if(typeof cfg.body === 'object') {
            if(isError(cfg.body)) {
                cfg.body = [
                    '```',
                    cfg.body.toString(),
                    cfg.body.stack,
                    '```'
                ].join('\n');
            } else {
                cfg.body = [
                    '```',
                    JSON.stringify(cfg.body, null, 4),
                    '```'
                ].join('\n');
            }
            
            cfg.markdown = true;
        }
        
        let params = {
            title: `${opts.titlePrefix || ''}${cfg.title}`,
            text: `${opts.bodyPrefix || ''}${cfg.body}${opts.bodyPostfix || ''}`,
            priority: cfg.priority || opts.priority || 'normal',
            tags: opts.tags || [],
            alert_type: cfg.type || 'info'
        };
        
        if(cfg.markdown || opts.markdown)
            params.text = `%%% \n ${params.text} \n %%%`;
        
        if(cfg.date && cfg.date instanceof Date)
            params.date_happened = Math.round(cfg.date.getTime() / 1000);
        
        if(cfg.host || opts.host)
            params.host = cfg.host || opts.host;
        
        if(cfg.tags && Array.isArray(cfg.tags))
            params.tags = params.tags.concat(cfg.tags);
        
        if(cfg.aggregationKey || opts.aggregationKey)
            params.aggregation_key = cfg.aggregationKey || opts.aggregationKey;
        
        if(cfg.sourceType || opts.sourceType)
            params.source_type_name = cfg.sourceType || opts.sourceType;
        
        return params;
    }
}

/**
 * Shortcut for creating a new instance of DataDogEvents.
 * 
 * @param {?Object} [options={}] Options to pass to DataDogEvents.
 * @returns {DataDogEvents}
 */
function dataDogEvents(options = {}) {
    return new DataDogEvents(options);
}

dataDogEvents.DataDogEvents = DataDogEvents;

module.exports = dataDogEvents;
